### PR TITLE
Avoid blocking on channel close on network thread

### PR DESCRIPTION
**Pull request by Tim-Brooks**
Currently when we close a channel in Netty4Utils.closeChannels we
block until the closing is complete. This introduces the possibility
that a network selector thread will block while waiting until a
separate network selector thread closes a channel.

For instance: T1 closes channel 1 (which is assigned to a T1 selector).
Channel 1's close listener executes the closing of the node. That
means that T1 now tries to close channel 2. However, channel 2 is
assigned to a selector that is running on T2. T1 now must wait until T2
closes that channel at some point in the future.

This commit addresses this by adding a boolean to `closeChannels`
indicating if we should block on close. We only set this boolean to true
if we are closing down the server channels at shutdown. This call is
never made from a network thread. When we call the closeChannels method
with that boolean set to false, we do not block on close.

### CODE REVIEW

**Comment by tbrooks8** [comment]
We have some assertions that check we never dereference our futures on network threads. But obviously those protections don't catch netty futures.

Another approach would be to not "wait" on the future completion when closing channels. Just close and move on. We could attach a listener to log on exception which we do now. If we went with that approach, we would might need a different code path at shutdown where we do wait on the channel being closed.

**Comment by s1monw** [comment]
what test triggered this behavior? I just removed this, IMO overengineered, behavior and we don't have any failures do we?

**Comment by s1monw** [comment]
@tbrooks8 can we maybe instead try to not block on the close future in the implementations and instead register each future in a concurrent set such that we can wait for them when we close the transport? to maintain the set we can also register a listener and clean it up once it's closed? does this all make sense?

**Comment by tbrooks8** [comment]
Yes. I can make some changes. I'll let you now when I'm ready for another review.

**Comment by tbrooks8** [comment]
@s1monw This is ready for another review. Instead of managing a set, I just added a boolean indicating if we wanted the close to be synchronous. This is valuable because at shutdown, we want the closing of the server sockets to be complete before the method returns.

In terms of the normal channels at shutdown, both the `Netty4Transport` and `NioTransport` independently ensure that all the channels are closed during shutdown (by blocking on futures).

File: core/src/main/java/org/elasticsearch/transport/TcpTransport.java [hunk 1]
```diff
@@ -1015,9 +1015,14 @@ protected void innerOnFailure(Exception e) {
     protected abstract Channel bind(String name, InetSocketAddress address) throws IOException;
 
     /**
-     * Closes all channels in this list
+     * Closes all channels in this list. If the synchronous boolean is set to true, the channels must be
+     * closed before the method returns. This boolean should never be called with true from a network
+     * thread.
+     *
+     * @param channels the channels to close
+     * @param synchronous whether the channels should be closed synchronously
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/transport/TcpTransport.java]: it's a matter of taste so I will leave it to you my suggestion is to call it `blocking` instead. but again up to you!

**Review by s1monw — APPROVED** [review]
great solution! LGTM

**Pull request merged**